File: lib/regression.cpp
#include <vector>
#include <numeric>
#include <math.h>

struct FitLine {
  double slope;
  double intercept;
  double r2;
};

FitLine regression(std::vector<double> y, std::vector<double> x) {
  const size_t n = y.size();
  const double sumX = std::accumulate(x.begin(), x.end(), 0.0);
  const double sumY = std::accumulate(y.begin(), y.end(), 0.0);
  const double sumXy = std::inner_product(x.begin(), x.end(), y.begin(), 0.0);
  const double sumXx = std::inner_product(x.begin(), x.end(), x.begin(), 0.0);
  const double sumYy = std::inner_product(y.begin(), y.end(), y.begin(), 0.0);

  FitLine result;
  result.slope = (n * sumXy - sumX * sumY) / (n * sumXx - sumX * sumX);
  result.intercept = (sumY - result.slope * sumX) / n;
  result.r2 = pow((n * sumXy - sumX * sumY) /
    sqrt((n * sumXx - sumX * sumX) * (n * sumYy - sumY * sumY)), 2);

  return result;
}

File: lib/regression.js
/* eslint-disable no-mixed-operators */

function linearRegression(y, x) {
  const n = y.length;
  let sumX = 0;
  let sumY = 0;
  let sumXy = 0;
  let sumXx = 0;
  let sumYy = 0;

  for (let i = 0; i < n; i++) {
    sumX += x[i];
    sumY += y[i];
    sumXy += (x[i] * y[i]);
    sumXx += (x[i] * x[i]);
    sumYy += (y[i] * y[i]);
  }

  const slope = (n * sumXy - sumX * sumY) / (n * sumXx - sumX * sumX);
  const intercept = (sumY - slope * sumX) / n;
  const r2 = ((n * sumXy - sumX * sumY) /
    Math.sqrt((n * sumXx - sumX * sumX) * (n * sumYy - sumY * sumY))) ** 2;

  return [slope, intercept, r2];
}

module.exports = linearRegression;
